maas-region/src/backups.py

```diff
@@ -487,17 +491,154 @@ def _on_restore_from_backup_action(self, event):
+        self.charm.unit.status = MaintenanceStatus("restoring backup")
 
-        # Step 1
-        logger.info("Step 1")
-        # ...
-        # ...
-        # Step N
-        logger.info("Step N")
+        self._run_restore(event, s3_parameters, backup_id)
 
+        self.charm.unit.status = ActiveStatus()
         event.set_results({"restore-status": "restore finished"})
 
+    def _run_restore(
+        self, event: ActionEvent, s3_parameters: dict[str, str], backup_id: str
+    ) -> None:
+        # Remove relation between maas-region and postgresql. This will stop the maas-region charm based on https://github.com/canonical/maas-charms/pull/511 .
+        if not self._relation_exists("maas-db"):
+            event.fail(
+                "PostgreSQL relation still exists, please run:\n"
+                "juju remove-relation maas-region postgresql\n"
+                "then retry this action"
+            )
+            return
+
+        # Ensure there are the same number of maas-region units as there are controllers in controllers.txt. i.e. 1 for non ha mode, or 3 for ha.
+        path = f"/backup/{backup_id}"
+        controllers_content = self._read_content_from_s3(f"{path}/controllers.txt", s3_parameters)
+        if not controllers_content:
+            event.fail("Restore failed: Could not fetch controllers list")
+            return
+
+        controllers = controllers_content.strip("\n").split("\n")
+        regions = relation.units if (relation := self.model.get_relation("maas-cluster")) else None
+        if regions is None:
+            event.fail("Restore failed: could not fetch MAAS regions list")
+            return
+
+        if len(controllers) != len(regions):
+            event.fail(
+                f"Restore failed: the number of MAAS-Region units ({len(regions)}) "
+                f"does not match the expected value from the backup ({len(controllers)})."
+            )
+
+        # For each region id in controllers.txt, add the region id to /var/snap/maas/common/maas/maas_id in a relevant maas-region unit.#
+        for region, controller in zip(regions, sorted(controllers)):
+            if self.model.name == region.name:
+                logger.debug(f"Assigning {controller} to {self.model.name}")
+                self._write_region_id(controller)
+
+        # Remove the existing image-storage i.e. sudo rm -rf /var/snap/maas/common/maas/image-storage
+        image_storage = SNAP_COMMON / "image-storage"
+        shutil.rmtree(image_storage, ignore_errors=True)
+        if image_storage.exists():
+            event.fail("Could not remove existing image-storage")
+            return
+
+        # Unpack image-storage.tar.gz into that directory, i.e. sudo tar -zxvf /image-archive.tar.gz -C /
+        ca_chain = s3_parameters.get("tls-ca-chain", [])
+        with tempfile.NamedTemporaryFile() if ca_chain else nullcontext() as ca_file:
+            if ca_file:
```

Comment:
the instantiation of the S3 client should be moved to download_and_unarchive_from_s3 to adhere to the logic of other s3 related functions.